ckan/logic/action/update.py

```diff
@@ -353,6 +357,113 @@ def package_update(context, data_dict):
 
     return output
 
+
+def package_revise(context, data_dict):
+    '''Revise a dataset (package) selectively with match, filter and
+    update parameters.
+
+    You must be authorized to edit the dataset and the groups that it belongs
+    to.
+
+    :param match: a dict containing "id" or "name" values of the dataset
+                  to update, all values provided must match the current
+                  dataset values or a ValidationError will be raised. e.g.
+                  ``{"name": "my-data", "resources": {["name": "big.csv"]}}``
+                  would abort if the my-data dataset's first resource name
+                  is not "big.csv".
+    :type match: dict
+    :param filter: a list of string patterns of fields to remove from the
+                   current dataset. e.g. ``["-resources__1"]`` would remove the
+                   second resource, ``["+title", "+resources", "-*"]`` would
+                   remove all fields at the dataset level except title and
+                   all resources (default: ``[]``)
+    :type filter: list of string patterns
+    :param update: a dict with values to update/create after filtering
+                   e.g. ``{"resources": [{"description": "file here"}]}`` would
+                   update the description for the first resource
+    :type update: dict
+    :param include: a list of string pattern of fields to include in response
+                    e.g. ``["-*"]`` to return nothing (default: ``[]`` all
+                    fields returned)
+    :type include: list of string patterns
+
+    match and update parameters may also be passed as path keys, e.g.
+    ``update__resource__1f9ab__description="file here"`` would set the
+    description of resource id starting with "1f9ab" to "file here".
```

Comment:
I think we should mention that you can use numeric indexes as well as ids on paths. And also the `__extend` option, which is undocumented AFAICT:

```suggestion
    match and update parameters may also be passed as path keys, using
    either the item numeric index or its unique id (with a minimum of 5 characters), e.g.
    ``update__resource__1f9ab__description="file here"`` would set the
    description of the resource with id starting with "1f9ab" to "file here", and
    ``update__resource__1__description="file here"`` would do the same
    on the second resource in the dataset.
    
    The ``extend`` key can also be used on the update parameter to add
    a new item to a list, e.g. ``update__resources__extend=[{"name": "new resource", "url": "https://example.com"}]`` 
    will add a new resource to the dataset with the provided ``name`` and ``url``.
```